submissions/kissik/data-structures/index.js

```diff
@@ -0,0 +1,110 @@
+const http = require('http');
+
+const fs = require('fs');
+
+const port = process.env.PORT || 3210;
+
+const list = {};
+
+list.data = {};
+
+list.push = data => {
+  list.data = {
+    item: data,
+    next: list.data
+  };
+};
+
+list.pop = () => {
+  const a = list.data.item ? list.data.item : 'none';
+  if (list.data) list.data = list.data.next;
+  return a;
+};
+
+list.print = msg => {
+  let a = list.data;
+  let str = '';
+  str += `\nThe ${msg}:\n`;
+  while (a && a.item) {
+    str += `${a.item} -> `;
+    a = a.next;
+  }
+  return str;
+};
+
+list.remove = data => {
+  let a = list.data;
+  let next;
+  if (a.next) next = a.next;
+  if (a.item === data) list.data = a.next;
+  while (next && next.item !== data) {
+    next = next.next;
+    a = a.next;
+  }
+  if (next && next.item === data) a.next = next.next;
+};
+
+http
+  .createServer((req, res) => {
+    if (req.url === '/') {
+      res.writeHead(200, { 'Content-Type': 'text/html' });
+      fs.createReadStream('pages/index.html').pipe(res);
+    } else if (req.url === '/css/theme.css') {
+      res.writeHead(200, { 'Content-Type': 'text/css' });
+      fs.createReadStream('css/theme.css').pipe(res);
+    } else if (req.url === '/img/binoculars.jpg') {
+      res.writeHead(200, { 'Content-Type': 'image/jpeg' });
+      fs.createReadStream('img/binoculars.jpg').pipe(res);
+    } else if (req.url === '/js/script.js') {
+      res.writeHead(200, { 'Content-Type': 'application/js' });
+      fs.createReadStream('js/script.js').pipe(res);
+    } else {
+      res.writeHead(404, { 'Content-Type': 'text/html' });
+      fs.createReadStream('pages/404.html').pipe(res);
+    }
+
+    let body = [];
+
+    req
+      .on('data', chunk => body.push(chunk))
+      .on('end', () => {
+        body = body.toString().split('&');
```

Comment:
You can use `URL.searchParams` for such things :)

[Please see](https://developer.mozilla.org/en-US/docs/Web/API/URL/searchParams)

Reply:
i use method="POST" in form tag (pages/index.html file)

Reply:
Then reading of body 100% must be moved to separate function and we you can use [Object#fromEntries](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/fromEntries).

```javascript
const entries = body
  .toString()
  .split("&")
  .map(bodyValue => bodyValue.split("="));

const json = Object.fromEntries(entries);
```

But anyway I think that this construction too big for such simple operation. What will be in `body` before `toString()`?

Reply:
The Object.fromEntries() function is currently not available in any version of Node.js (

Reply:
"What will be in body before toString()?"

`application/x-www-form-urlencoded`: the keys and values are encoded in key-value tuples separated by `&`, with a `=` between the key and the value

Reply:
> The Object.fromEntries() function is currently not available in any version of Node.js (

Are you sure? I see that it should work from [12+](https://v8.dev/features/object-fromentries)

![image](https://user-images.githubusercontent.com/2997359/68762254-7c27f380-061e-11ea-912f-304734418556.png)

I have used 12.13.0, current LTS version. To get latest features, you can use 13+

Also, if you can't keep several versions of node on your machine, you can use [nvm](https://github.com/nvm-sh/nvm)